polkadot/primitives/src/vstaging/mod.rs

```diff
@@ -98,3 +116,873 @@ impl<BlockNumber: Default + From<u32>> Default for SchedulerParams<BlockNumber>
+	pub disputes: MultiDisputeStatementSet,
+	/// The parent block header. Used for checking state proofs.
+	pub parent_header: HDR,
+}
+
+impl<H> BackedCandidate<H> {
+	/// Constructor
+	pub fn new(
+		candidate: CommittedCandidateReceiptV2<H>,
+		validity_votes: Vec<ValidityAttestation>,
+		validator_indices: BitVec<u8, bitvec::order::Lsb0>,
+		core_index: Option<CoreIndex>,
+	) -> Self {
+		let mut instance = Self { candidate, validity_votes, validator_indices };
+		if let Some(core_index) = core_index {
+			instance.inject_core_index(core_index);
+		}
+		instance
+	}
+
+	/// Get a reference to the committed candidate receipt of the candidate.
+	pub fn candidate(&self) -> &CommittedCandidateReceiptV2<H> {
+		&self.candidate
+	}
+
+	/// Get a reference to the descriptor of the candidate.
+	pub fn descriptor(&self) -> &CandidateDescriptorV2<H> {
+		&self.candidate.descriptor
+	}
+
+	/// Get a reference to the validity votes of the candidate.
+	pub fn validity_votes(&self) -> &[ValidityAttestation] {
+		&self.validity_votes
+	}
+
+	/// Get a mutable reference to validity votes of the para.
+	pub fn validity_votes_mut(&mut self) -> &mut Vec<ValidityAttestation> {
+		&mut self.validity_votes
+	}
+
+	/// Compute this candidate's hash.
+	pub fn hash(&self) -> CandidateHash
+	where
+		H: Clone + Encode,
+	{
+		self.candidate.to_plain().hash()
+	}
+
+	/// Get this candidate's receipt.
+	pub fn receipt(&self) -> CandidateReceiptV2<H>
+	where
+		H: Clone,
+	{
+		self.candidate.to_plain()
+	}
+
+	/// Get a copy of the validator indices and the assumed core index, if any.
+	pub fn validator_indices_and_core_index(
+		&self,
+		core_index_enabled: bool,
+	) -> (&BitSlice<u8, bitvec::order::Lsb0>, Option<CoreIndex>) {
+		// This flag tells us if the block producers must enable Elastic Scaling MVP hack.
+		// It extends `BackedCandidate::validity_indices` to store a 8 bit core index.
+		if core_index_enabled {
+			let core_idx_offset = self.validator_indices.len().saturating_sub(8);
+			if core_idx_offset > 0 {
+				let (validator_indices_slice, core_idx_slice) =
+					self.validator_indices.split_at(core_idx_offset);
+				return (
+					validator_indices_slice,
+					Some(CoreIndex(core_idx_slice.load::<u8>() as u32)),
+				);
+			}
+		}
+
+		(&self.validator_indices, None)
+	}
+
+	/// Inject a core index in the validator_indices bitvec.
+	fn inject_core_index(&mut self, core_index: CoreIndex) {
+		let core_index_to_inject: BitVec<u8, bitvec::order::Lsb0> =
+			BitVec::from_vec(vec![core_index.0 as u8]);
+		self.validator_indices.extend(core_index_to_inject);
+	}
+
+	/// Update the validator indices and core index in the candidate.
+	pub fn set_validator_indices_and_core_index(
+		&mut self,
+		new_indices: BitVec<u8, bitvec::order::Lsb0>,
+		maybe_core_index: Option<CoreIndex>,
+	) {
+		self.validator_indices = new_indices;
+
+		if let Some(core_index) = maybe_core_index {
+			self.inject_core_index(core_index);
+		}
+	}
```

Comment:
These methods will need to change, as we now have an additional way of retrieving and setting the core index

Reply:
Yes, we'd have to make a decision if we still want to keep the injected core index as backwards compatibility. This should be done in the PR that adds runtime support for the new receipts.

Reply:
(I will update the RFC with my proposal)